Solve async challenges

ezrillex opened the pull request:
Thanks in advance for the feedback.

ezrillex committed:
Solve challenges

ezrillex committed:
Develop promisified solution. Now it makes sense  🤯🤯🤯

ezrillex committed:
ESlinted semi

erickvh commented on .idea/vcs.xml:
We can avoid to push these files when it is not necessary for the project.

ezrillex committed:
Implement feedback commonjs and enforce with eslint

erickvh commented on package.json:
```diff
@@ -0,0 +1,16 @@
+  "name": "js-async-nerdery-challenges",
+  "version": "1.0.0",
+  "description": "This repo provides a series of exercises to put your js async understanding into practice.",
+  "main": "index.js",
```
Nit: it is not necessary to define an entry point if it doesn't exist the file

erickvh commented on eslint.config.mjs:
I think we need to have more consistency if we're using commonjs insted of esmodules.

erickvh commented on eslint.config.mjs:
```diff
@@ -0,0 +1,13 @@
+import globals from "globals";
```
A module import here, should it be in this way?

erickvh commented on eslint.config.mjs:
```diff
@@ -0,0 +1,13 @@
+/** @type {import('eslint').Linter.Config[]} */
+export default [
+  {languageOptions: {
+    globals: globals.browser ,
```
I think we can re-think if this global is the one that we currently want to use

erickvh commented on 1-callbacks/solution.js:
```diff
@@ -31,17 +31,54 @@ node solution.js name1 name2 name3
+    };
 
     // iterate the names array and validate them with the method
+    for (let index = 0; index < names.length; index++) {
+        const name = names[index];
+        validateUser(name, myCallback);
+    }
```
You can use a foreach instead and directly use the name value on each iteration

erickvh commented on 1-callbacks/solution.js:
```diff
@@ -31,17 +31,54 @@ node solution.js name1 name2 name3
+        names.push(...process.argv.slice(2));
+    }
+    
+    const okNames = [];
+    const errNames = [];
```
You can improve readability by using names like.

`authorizedUsers`
`unAuthorizedUsers`

erickvh commented on 1-callbacks/solution.js:
```diff
@@ -31,17 +31,54 @@ node solution.js name1 name2 name3
+    let cbCount = 0 ; // console results are shown before all cbs finish, so I will add logic to avoid showing until last cb
+
+    // YOUR SOLUTION GOES HERE
+    const myCallback = (...params)=>{ // expand because if is in list it sends 2
```
The callback name could be improve

erickvh replied:
You can also define the callback function outside of this function to keep your implementation cleaner.

erickvh commented on 1-callbacks/solution-promisified.js:
```diff
@@ -0,0 +1,78 @@
+
+    const validationPromises = [];
+    // iterate the names array and validate them with the method
+    for (let index = 0; index < names.length; index++) {
+        // so callback that we used to pass is now passed by the promisified wrapper.
+        // because my callback was just receivieng and pushing to results array the results.
+        // this is done automatically by promisified callback. thus only passing names. 
+        validationPromises.push(promValidateUser(names[index]))  ;
+    }
+    const data = await Promise.allSettled(validationPromises);
```
This whole logic could be handle by a map and this map could be handle in the `promise.allSettled()` method.

erickvh commented on 1-callbacks/solution-promisified.js:
```diff
@@ -0,0 +1,78 @@
+    // log the final result
+    function showResults(){
+        console.log("Success");
+        data.filter(item=>item.status === "fulfilled").forEach(item => console.log(`Id: ${item.value.id}\nName: ${item.value.name}`));
+        console.log("\nFailure");
+        data.filter(item=>item.status === "rejected").forEach(item => console.log(item.reason.message));
```
You can use a reduce to get both groups full filled and failed promises so in that way you will do just an iteration in the array

ezrillex replied:
Now I really understand how to reduce! 
![image](https://github.com/user-attachments/assets/7ac73d47-f506-418e-83df-71dd2b202a40)

erickvh commented on 1-callbacks/solution.js:
```diff
@@ -31,17 +31,54 @@ node solution.js name1 name2 name3
     // log the final result
+    function showResults(){
+        console.log("Success");
+        okNames.forEach(item => console.log(`Id: ${item.id}\nName: ${item.name}`));
+        console.log("\nFailure");
+        errNames.forEach(item => console.log(item));
```
You can group them using reduce

erickvh commented on 1-callbacks/solution-promisified.js:
```diff
@@ -0,0 +1,78 @@
+    // Assuming promisifying the whole funciton is valid. And this is following the "error first pattern". 
+    // I think I understood wrong, I thought initially the callback was the only one to be wrapped in a promise. 
+    // But this makes more sense since I pass it a callback and promisify wrapper adds a callback to receive the value.
+    const promValidateUser = util.promisify(validateUser);
```
```suggestion
    const promisifiedValidateUsed = util.promisify(validateUser);
```

erickvh commented on 2-promises/solution.js:
```diff
@@ -24,16 +24,37 @@ const id = yourRandomMethod() //third run
+    })
+    
     // You log the fullname, or error, here
+    .then(() =>{
+        console.log(`The full name is ${first} ${last}.`);
+    })
```
I think you can improve it, by removing this nested since the previous promise returns a undefined value

erickvh commented on 2-promises/solution.js:
```diff
@@ -24,16 +24,37 @@ const id = yourRandomMethod() //third run
 
     // You call the firstname method
-
+    .then((lastname)=>{
+        last = lastname;
+        return firstnames(lastname);
+    })
+    .then(firstname=>{
+        first = firstname;
+    })
```
I think we don't need to use a variable for last and first name here, you can rely just in the nested promises.

I recommend you to use the method then with first name method

ezrillex replied:
// You call the lastnames method with your id
    const lastnames_promise = lastnames_function(random);

    // You call the firstname method
    const firstnames_promise = lastnames_promise.then((lastname)=>{
        return firstnames_function(lastname)
    });

    // You log the fullname, or error, here
    const full_name = Promise.all([firstnames_promise, lastnames_promise]);

    full_name.then(results=>{
        console.log('The full name is' , results.reduce((acc, curr) => {acc = acc + curr + ' '; return acc},""))
    })
    .catch(err=>{
        console.log("There was an error. Message = ", err.message);
    })

ezrillex replied:
This is how it ended up, not sure if aligned with what you had in mind but this is not using any external variable other than the promises themselves. For the passing an object approach the constraint of not modifying the functions needs to be lifted but I have it in mind as another way.

erickvh commented on 2-promises/solution.js:
```diff
@@ -24,16 +24,37 @@ const id = yourRandomMethod() //third run
 7. log the resultant fullname, or the error, at the end
 */
 
+const firstnames = require("./firstnames.js");
+const lastnames = require("./lastnames.js");
```
Since they're default methods feel free to rename it as better as you can

erickvh commented on 3-async-await/solution.js:
```diff
@@ -19,17 +19,83 @@ Example:
-function solution() {
-    // YOUR SOLUTION GOES HERE
+async function solution() {
+    // is this a bad practice? declaring but no initalization of variables that will eventually have a value in all logic branches?
```
It really depends on the context of your method, but in this case is totally valid since you're going to use each variable.

In some cases it is not a good idea to initialize all of the variables due to internal represents so space reserved to do something.

For example if you're not going to use a variable because the method has an error abruptly it can be an example of not assigning a variable.

erickvh commented on 3-async-await/solution.js:
```diff
@@ -19,17 +19,83 @@ Example:
 
     // You use Promise.all() here
+    try {
+        allResult = await Promise.all([prices(rng), products(rng)]);
```
in this case it is better to destructure the array since you already know which is index 0 and index 1
```suggestion
       const [price, product] = await Promise.all([prices(rng), products(rng)]);
```

erickvh commented on 3-async-await/solution.js:
```diff
@@ -19,17 +19,83 @@ Example:
+    try {
+        settledResult = await Promise.allSettled([products(rng), prices(rng)]);
+    }
+    catch (error) {
+        settledResult = error;
+    }
```
We don't need to handle an error here since promise all setlled already catch them by you internally.

erickvh commented on 3-async-await/solution.js:
```diff
@@ -19,17 +19,83 @@ Example:
+        if (item.status === 'rejected') {
+            return ({
+                status: item.status,
+                name: index === 1 ? "Price" : "Product",
```
This is duplicating logic, that can be address by destructuring

ezrillex committed:
Update 1-callbacks/solution-promisified.js

Co-authored-by: Erick Ventura <erick94111@gmail.com>

ezrillex committed:
remove non essential files

ezrillex committed:
Remove unused entrypoint

ezrillex committed:
Regenerate eslint config file.

ezrillex committed:
resolve feedback of solution 1

avoid declaring functions in functions, moved independent parts to top level.  Simplified loop for validation of names.

ezrillex committed:
chore/cleanup repo from ide files

ezrillex committed:
Merge branch 'solution/ezra-abarca' of https://github.com/ezrillex/JS-async-Nerdery-Challenges into solution/ezra-abarca

ezrillex committed:
feat/implement reducer in promises and display

ezrillex committed:
rewrite/better naming and reworked promise nesting

ezrillex committed:
refactor/ reduce redundancies and reorganized

ezrillex committed:
chore/ cleanup extra files

ezrillex committed:
chore/ cleanup extra files

ezrillex committed:
chore/ cleanup extra files

ezrillex committed:
chore/ cleanup extra configuration

ezrillex committed:
Revert "chore/ cleanup extra configuration"

This reverts commit e47918960949dec01fe9fc31492643bf0f02c85e.
